docs/docs/vm-fundamentals/builtins.md

```diff
@@ -4,4 +4,143 @@ sidebar_position: 2
+
+### What is builtins in cairo?
+
+Builtins are predefined optimized low-level execution units which are added to the Cairo CPU board to perform predefined computations which are expensive to perform in Cairo.
+
+### Builtins and Cairo memory
+
+Communication between the CPU and built-in functionalities occurs via memory-mapped I/O. Each builtin is allocated a contiguous memory region and enforces specific constraints on the data residing within that area. The Pedersen builtin is a good example to explain this.
+
+Pedersen builtin establishes that:
+
+```
+[p + 2] = hash([p + 0], [p + 1])
+[p + 5] = hash([p + 3], [p + 4])
+[p + 8] = hash([p + 6], [p + 7])
+```
+
+The Cairo code may read or write from this memory cells to “invoke” the builtin. The following code verifies that `hash(x, y) == z`
+
+```
+// Write the value of x to [p + 0].
+x=[p]; // Write the value of y to [p + 1].
+y=[p + 1]; // The builtin makes that [p + 2] == hash([p + 0], [p + 1]).
+z=[p + 2];
+```
+
+Cairo memory immutability requires a careful usage of builtin instances because memory locations like `[p + 0]`, `[p + 1]`, and `[p + 2]` are used for a single hash computation, they cannot be reused for subsequent calculations and needs to tracking an unused memory location pointer (`hash_ptr`).
+
+By convention, the functions that utilizing builtins receive a pointer to the unused memory location as an argument and return an updated pointer reflecting the next available memory slot. For example:
+
+```
+func hash2(hash_ptr: felt*, x, y) -> (hash_ptr: felt*, z: felt) {
+  // Invoke the hash function.
+  x = [hash_ptr];
+  y = [hash_ptr + 1];
+  // Update pointer (increment by 3) and return result.
+  return (hash_ptr=hash_ptr + 3, z=[hash_ptr + 2]);
+}
+```
+
+On the other hand, Starkware Cairo library provides typed references for improved readability. Here's the code rewritten with `HashBuiltin` type like this example:
+
+```
+from starkware.cairo.common.cairo_builtins import HashBuiltin
+
+func hash2(hash_ptr: HashBuiltin*, x, y) -> (hash_ptr: HashBuiltin*, z: felt) {
+  let hash = hash_ptr;
+  // Invoke the hash function using typed access.
+  hash.x = x;
+  hash.y = y;
+  // Update pointer and return result.
+  return (hash_ptr=hash_ptr + HashBuiltin.SIZE, z=hash.result);
+}
+```
+
+### Some specific builtins in Cairo VM
+
+- **output**
+    
+    The output builtin, accessed with a pointer to type felt, is used for writing program outputs.
+    
+    ```rust
+    %builtins output 
+    //allows the program to write this output to a designated memory location.
+    from starkware.cairo.common.serialize import serialize_word
+    func main{output_ptr: felt*}() { 
+    //the output here indicate that parameter is a pointer to a memory location of 
+    //type felt*. The location is used to store the program output 
+        serialize_word(1234);
+        serialize_word(4321);
+        return ();
+    }
+    ```
+    
+- **pedersen**
```

Comment:
Let's make all the remaining builtins described (`pedersen`,  `range_check`, `ecdsa` & `bitwise`) to be explained just as `output` is described

Reply:
meaning it's not cairo code. I cannot run the code in this examples because it won't compile